### PR TITLE
Add engine subsystem abstractions

### PR DESCRIPTION
**Pull request by iamyulong**
This PR introduces `ResourceSystem` and `ComponentSystem` abstractions.

It follows https://github.com/radixdlt/radixdlt-scrypto/pull/162

### CODE REVIEW

**Comment by iamyulong** [comment]
Rebased this PR on https://github.com/radixdlt/radixdlt-scrypto/pull/162

File: scrypto/src/resource/system.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
+use crate::engine::{api::*, call_engine};
+use crate::resource::*;
+use crate::rust::collections::HashMap;
+use crate::rust::string::String;
+
+/// Represents the Radix Engine resource subsystem.
```
Comment (talekhinezh) [inline scrypto/src/resource/system.rs:6]: Rather than subsystem perhaps call it a driver?

**Comment by iamyulong** [comment]
Apologies for the accidental auto-merge by https://github.com/radixdlt/radixdlt-scrypto/pull/161.

I will follow up if there is further comments.

File: scrypto/src/resource/system.rs (hunk 1 above)
Comment (iamyulong) [reply scrypto/src/resource/system.rs:6]: We may call it "driver" (as loadable module) in Radix Engine. However, this is a developer facing API; driver is less appropriate IMO.

**Pull request merged**